controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,285 @@
+			labels["internal.gatekeeper.sh/constraint-name"])
+	}
+
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for kind, names := range constraintKindNames {
+		constraintGVR := schema.GroupVersionResource{
+			Group:    "constraints.gatekeeper.sh",
+			Version:  "v1beta1",
+			Resource: strings.ToLower(kind),
+		}
+		for _, name := range names {
+			constraint, err := r.DynamicClient.Resource(constraintGVR).Get(ctx, name, metav1.GetOptions{})
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			constraintKinds, _, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			r.addSyncOnlys(constraintKinds, &syncOnlys)
+		}
+	}
+
+	deleteDuplicateSyncOnly(&syncOnlys)
+	config.Spec.Sync.SyncOnly = syncOnlys
+	err = r.Update(ctx, config, &client.UpdateOptions{})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func deleteDuplicateSyncOnly(syncOnlys *[]v1alpha1.SyncOnlyEntry) {
+	unique := []v1alpha1.SyncOnlyEntry{}
+
+	for _, sync := range *syncOnlys {
+		isUniq := true
+
+		for _, uniq := range unique {
+			if uniq.Group == sync.Group &&
+				uniq.Version == sync.Version &&
+				uniq.Kind == sync.Kind {
+				isUniq = false
+
+				break
+			}
+		}
+
+		if isUniq {
+			unique = append(unique, sync)
+		}
+	}
+
+	*syncOnlys = unique
+}
+
+func (r *ConstraintStatusReconciler) addSyncOnlys(constraintKinds []interface{}, syncOnlys *[]v1alpha1.SyncOnlyEntry) {
+	for _, kind := range constraintKinds {
+		newKind := kind.(map[string]interface{})
+		apiGroups := newKind["apiGroups"].([]interface{})
+		kindsInKinds := newKind["kinds"].([]interface{})
+
+		for _, apiGroup := range apiGroups {
+			for _, kindkind := range kindsInKinds {
+				version := r.getApiVersion(kindkind.(string), apiGroup.(string))
+
+				*syncOnlys = append(*syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kindkind.(string),
+				})
+			}
+		}
+	}
+}
+
+func (r *ConstraintStatusReconciler) getApiVersion(kind string, apiGroup string) string {
+	// cool time(1 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*1).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return ""
+		}
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison := strings.Split(resc.GroupVersion, "/")
+		var group string
+		var version string
+		// consider groupversion == v1 or groupversion == app1/v1
+		if len(groupVerison) == 2 {
+			group = groupVerison[0]
+			version = groupVerison[1]
+		} else {
+			version = groupVerison[0]
+		}
+
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version
+			}
+		}
+	}
+
+	return ""
+}
+
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
```

Comment:
Need discovery because users only provide api-group and kind name but config sync only needs version.